NPS_generation/calculate_outcomes.py

```diff
@@ -99,383 +60,425 @@ def pct_stereocentres(mol):
+                org_fps = get_ecfp6_fingerprints(org_mols)
+                external_div = external_diversity(gen_fps, org_fps)
+                res = res._append(pd.DataFrame({
                     'input_file': sampled_file,
-                    'outcome': ['KL divergence, NP score',
-                                'Jensen-Shannon distance, NP score',
-                                'Wasserstein distance, NP score'],
-                    'value': [kl_NP, jsd_NP, emd_NP] }))
-
-            ## outcome 15: K-L divergence of % sp3 carbons
-            gen_sp3 = [Lipinski.FractionCSP3(mol) for mol in gen_mols]
-            kl_sp3 = continuous_KL(gen_sp3, org_sp3)
-            jsd_sp3 = continuous_JSD(gen_sp3, org_sp3)
-            emd_sp3 = continuous_EMD(gen_sp3, org_sp3)
-            res = res.append(pd.DataFrame({
+                    'outcome': 'External diversity',
+                    'value': [external_div] }))
+
+                ## also, summarize using nearest-neighbor instead of mean
+                internal_= internal_nn(gen_fps)
+                external_ = external_nn(gen_fps, org_fps)
+                res = res._append(pd.DataFrame({
                     'input_file': sampled_file,
-                    'outcome': ['KL divergence, % sp3 carbons',
-                                'Jensen-Shannon distance, % sp3 carbons',
-                                'Wasserstein distance, % sp3 carbons'],
-                    'value': [kl_sp3, jsd_sp3, emd_sp3] }))
-
-            ## outcome 16: K-L divergence of % rotatable bonds
-            gen_rot = [pct_rotatable_bonds(mol) for mol in gen_mols]
-            kl_rot = continuous_KL(gen_rot, org_rot)
-            jsd_rot = continuous_JSD(gen_rot, org_rot)
-            emd_rot = continuous_EMD(gen_rot, org_rot)
-            res = res.append(pd.DataFrame({
+                    'outcome': ['External nearest-neighbor Tc',
+                                'Internal nearest-neighbor Tc'],
+                    'value': [external_, internal_] }))
+
+                ## outcome 12: K-L divergence of number of rings
+                gen_rings1 = [Lipinski.RingCount(mol) for mol in gen_mols]
+                gen_rings2 = [Lipinski.NumAliphaticRings(mol) for mol in gen_mols]
+                gen_rings3 = [Lipinski.NumAliphaticRings(mol) for mol in gen_mols]
+                kl_rings1 = discrete_KL(gen_rings1, org_rings1)
+                kl_rings2 = discrete_KL(gen_rings2, org_rings2)
+                kl_rings3 = discrete_KL(gen_rings3, org_rings3)
+                jsd_rings1 = discrete_JSD(gen_rings1, org_rings1)
+                jsd_rings2 = discrete_JSD(gen_rings2, org_rings2)
+                jsd_rings3 = discrete_JSD(gen_rings3, org_rings3)
+                emd_rings1 = discrete_EMD(gen_rings1, org_rings1)
+                emd_rings2 = discrete_EMD(gen_rings2, org_rings2)
+                emd_rings3 = discrete_EMD(gen_rings3, org_rings3)
+                res = res._append(pd.DataFrame({
```

Comment:
We should use `res = res.concat([res, pd.DataFrame...], axis=1)

Reply:
Changed every occurrence of this. The correct code is `res = pd.concat([res, pd.DataFrame...], axis=1)`

Reply:
Looks like it should be `axis=0` instead. Sorry about that. Let me fix and push.